### PR TITLE
Fix collateral balancing when building transaction

### PR DESCRIPTION
**Pull request by carbolymer**
# Changelog

```yaml
- description: |
    Fix collateral balancing when building transaction
# uncomment types applicable to the change:
  type:
  # - feature        # introduces a new feature
  # - breaking       # the API has changed in a breaking way
   - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - refactoring    # QoL changes
   - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

This PR fixes: 
- https://github.com/IntersectMBO/cardano-cli/issues/754

This PR consist of two commits:
* "Refactor: Remove Test.Cardano.Api.Typed and Test.Golden.Cardano.Api.Typed modules" - reorganisation of the test module - no new code was added there.
* "Fix collateral balancing when transaction building, add unit and property tests" - actual fix

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Self-reviewed the diff

### CODE REVIEW

File: cardano-api/internal/Cardano/Api/Fees.hs [hunk 1]
```diff
@@ -1295,49 +1297,52 @@ calcReturnAndTotalCollateral
...
-                ( TxOut
-                    cAddr
-                    (lovelaceToTxOutValue (babbageEraOnwardsToShelleyBasedEra retColSup) returnCollateral)
-                    TxOutDatumNone
-                    ReferenceScriptNone
-                )
-            , totalCollateral
-            )
-          else (TxReturnCollateralNone, TxTotalCollateralNone)
+calcReturnAndTotalCollateral w fee pp' TxInsCollateral{} txReturnCollateral txTotalCollateral cAddr totalAvailableCollateral = babbageEraOnwardsConstraints w $ do
```
Comment (Jimbo4350) [inline cardano-api/internal/Cardano/Api/Fees.hs:1304]: Does this function exist in `cardano-ledger`?  Or any function that can assist in this calculation? The more we can call from ledger the better.

File: cardano-api/test/cardano-api-test/Test/Cardano/Api/Transaction/Autobalance.hs [hunk 2]
```diff
@@ -15,35 +15,50 @@ module Test.Cardano.Api.Transaction.Autobalance
...
+import qualified Hedgehog.Gen as Gen
 import           Test.Tasty (TestTree, testGroup)
 import           Test.Tasty.Hedgehog (testProperty)
 
 -- | Test that the fee is the same when spending minted asset manually or when autobalancing it
 prop_make_transaction_body_autobalance_return_correct_fee_for_multi_asset :: Property
 prop_make_transaction_body_autobalance_return_correct_fee_for_multi_asset = H.propertyOnce $ do
-  let sbe = ShelleyBasedEraConway
+  let ceo = ConwayEraOnwardsConway
+      beo = conwayEraOnwardsToBabbageEraOnwards ceo
```
Comment (Jimbo4350) [inline cardano-api/test/cardano-api-test/Test/Cardano/Api/Transaction/Autobalance.hs:59]: Another good example as to why we have to ditch the `Eon`s.

**Review by Jimbo4350 — APPROVED** [review]
Great work, I can tell this was annoying to debug. Just one comment regarding the commit message.

File: cardano-api/internal/Cardano/Api/Fees.hs (hunk 1 above)
Comment (carbolymer) [reply cardano-api/internal/Cardano/Api/Fees.hs:1304]: Something similar is implemented here, but it's a validation, not a calculation, so we can't just reuse it. https://github.com/IntersectMBO/cardano-ledger/blob/master/eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs#L307

**Pull request merged**